# Bias and Fairness Assessment of Perspective API Model

## Part 1: Setting Up

Importing the needed libraries and modules

In [1]:
import pandas as pd 
import numpy as np
import nltk 
from nltk.corpus import stopwords 
from collections import Counter

pd.options.mode.chained_assignment = None  # default='warn'
from sklearn.metrics import accuracy_score,confusion_matrix
from sklearn.metrics import classification_report
from sklearn.ensemble import GradientBoostingClassifier

## Part 2: Exploring The Sample Dataset

**Load Test Data Set** `Sample_labeled_data.csv`

In [2]:
df = pd.read_csv("Sample_labeled_data.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55252 entries, 0 to 55251
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Unnamed: 0    55252 non-null  int64 
 1   id            55252 non-null  object
 2   comment_text  55246 non-null  object
 3   toxic         55252 non-null  object
dtypes: int64(1), object(3)
memory usage: 1.7+ MB


In [3]:
df.head()

Unnamed: 0                id  \
0           5  0001ea8717f6de06   
1           7  000247e83dcc1211   
2          11  0002f87b16116a7f   
3          13  0003e1cccfd5a40a   
4          14  00059ace3e3e9a53   

                                        comment_text toxic  
0  Thank you for understanding I think very highl...    no  
1                     Dear god this site is horrible    no  
2   Somebody will invariably try to add Religion ...    no  
3    It says it right there that it IS a type The...    no  
4     Before adding a new product to the list mak...    no

### Cleaning Data
#### Remove unwanted columns : `'Unnamed: 0'` & `'id'`

In [4]:
df.drop(columns=['Unnamed: 0', 'id'], inplace=True)

df.info()
#df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55252 entries, 0 to 55251
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   comment_text  55246 non-null  object
 1   toxic         55252 non-null  object
dtypes: object(2)
memory usage: 863.4+ KB


#### Extract Non-Latinized Comments

After looking through the data set, I noticed there were some comments written in Chinese or Arabic characters. So, to clean the data, we will extract all rows in which the comment_text has non-latin characters. 

In [5]:
import re

df.fillna('', inplace=True)

# Remove non-Latin characters from the comment_text column
df['comment_text'] = df['comment_text'].apply(lambda x: re.sub(r'[^\x00-\x7F]+', ' ', x))

# Perform language detection and filter out non-English comments as before
mask = df['comment_text'].str.strip().eq('')

# Use the boolean mask to select the rows you want to remove, and use drop() method to remove them
df.drop(df[mask].index, inplace=True)

### Creating New Feature Columns
#### Directed Towards, Who?

I labelled the comments as directed towards 'male', 'female', or 'non-gendered' based on the most commonly used pronoun found within the comment text. 

'Male': He/Him/His

'Female': She/Her/Hers

'Non-gendered': They/Them/Their/You/Your/yours

If there were no pronouns found within the comment, I dropped the row from the dataframe.

In [6]:
def directed_towards(comment):
    pronouns = re.findall(r'\b(?:she|her|hers|he|him|his|they|them|theirs|you|your|yours|yourself|i|we|me|mine|myself)\b', comment.lower())
    if not pronouns:
        return 'none'
    pronoun_counts = Counter(pronouns)
    most_common_pronoun = pronoun_counts.most_common(1)[0][0]
    if most_common_pronoun in ('she', 'her', 'hers'):
        return 'female'
    elif most_common_pronoun in ('he', 'him', 'his'):
        return 'male'
    elif most_common_pronoun in ('they', 'them', 'theirs', 'you', 'your', 'yours', 'yourself', 'i', 'we', 'me', 'mine', 'myself'):
        return 'non-gendered'

# Apply the function to each comment in the dataframe
df['directed_towards'] = df['comment_text'].apply(directed_towards)

We will drop all the rows in which no pronoun is found.

In [7]:
df.drop(df[df['directed_towards'] == 'none'].index, inplace = True)
df.head()

comment_text toxic directed_towards
0   Thank you for understanding I think very highl...    no     non-gendered
2    Somebody will invariably try to add Religion ...    no     non-gendered
3     It says it right there that it IS a type The...    no     non-gendered
9   Please stop If you continue to vandalize Wikip...    no     non-gendered
10  RedSlash cut it short If you have sources stat...    no     non-gendered

### Load to CSV
Load the cleaned dataframe into a new csv file.

In [8]:
df.to_csv("Cleaned_Sample_Data.csv")

### Creating New DF
#### DF of Comments Labelled 'Toxic'
Create a new dataframe of all the comments in which the toxic label was assigned by manual reviewers

In [9]:
toxic_df = df[df["toxic"]=='yes']
toxic_df.to_csv('Toxic_Comments.csv')
toxic_df.head()

comment_text toxic directed_towards
34   How dare you vandalize that page about the HMS...   yes     non-gendered
36   No he is an arrogant self serving immature idi...   yes             male
81   Eek but shes cute in an earthy kind of way Can...   yes     non-gendered
114  we hate america and we are going to bomb the s...   yes     non-gendered
117  Moi Ego I am mortified that you could say such...   yes     non-gendered

### Finding Most Frequent Words

Converting text to word-tokenizing (i.e.creating individual strings of each word of a comment and housed in a list)

In [10]:
def frequent_nouns(df):
    comments = df['comment_text'].str.lower().tolist()
    tokens = [nltk.word_tokenize(comment) for comment in comments]
    
    all_tokens = []
    
    for token_list in tokens:
        all_tokens.extend(token_list)
    
    text = nltk.Text(all_tokens)

    #print(all_tokens)
    
    nouns = []
    
    for comment in comments:
        tokens = nltk.word_tokenize(comment)
        pos_tags = nltk.pos_tag(tokens)
        for word, pos in pos_tags:
            if pos.startswith('N'):
                nouns.append(word)

    # Print the most common nouns
    text = nltk.Text(nouns)
    return text.vocab().most_common(10)

#### Most Frequent Words In Toxic Comments for each Pronoun category

In [11]:
#Toxic Comment Directed Towards Female Individual
toxic_female = toxic_df[(toxic_df['directed_towards'] == 'female')]

#Toxic Comment Directed Towards Male Individual
toxic_male = toxic_df[(toxic_df['directed_towards'] == 'male')]
#df[(df['toxic'] == 'yes') & ((df['directed_towards'] == 'male'))]

#Toxic Comment Directed Towards Non-gendered Individual/s
toxic_non_gendered = toxic_df[(toxic_df['directed_towards'] == 'non-gendered')]

In [12]:
print('Directed Towards Female: ', frequent_nouns(toxic_female), '\n\n', 
      'Directed Towards Male: ', frequent_nouns(toxic_male),'\n\n'
      'Directed Towards Non-gendered: ', frequent_nouns(toxic_non_gendered),'\n\n')


Directed Towards Female:  [('bitch', 76), ('ok', 67), ('lo', 22), ('i', 15), ('sex', 15), ('woman', 12), ('man', 9), ('girl', 8), ('hell', 7), ('people', 6)] 

 Directed Towards Male:  [('i', 26), ('penis', 24), ('man', 21), ('hes', 19), ('people', 16), ('shit', 14), ('guy', 12), ('article', 12), ('sex', 12), ('page', 11)] 

Directed Towards Non-gendered:  [('i', 1246), ('youi', 880), ('hate', 835), ('vandal', 457), ('traitor', 447), ('c', 362), ('fgtyou', 219), ('people', 215), ('r', 195), ('k', 189)] 




**Insights** : 

In the top 10 words of toxic comments, those directed towards 'Female' and 'Male' individuals have words that are crude, refer to sexualized body parts, or are insults compared those directed towards non-gendered individuals.

- Female: 'bitch', 'sex', 'hell'
- Male: 'penis', 'sex', 'shit'

Comments that were considered 'non-gendered' held no sexualized and deeply offensive words except 'hate' and 'traitor'. 

There is similarity in the most frequent words between comments containing male and female pronouns. In the next section I will dive deeper to find if the Prospective API has any bias when labelling between the two. 

# Part 3: Hypotheses, Design, and Performing Tests

##### In this section, I will document and test the Perspective API model for bias. 


### 3.0 Setting a Threshold

**Model Threshhold**
In order to create a decision for the classification of user's perception of a comment to be toxic, a value, or threshold, must be created.  

I imported the Google API Client, utilized an API Key, and selected comments within the sample_labeled_data dataframe ('df') to retrieve the relevant Perspective API toxicity score, which I stored in a new sample_dataset.

In [13]:
from googleapiclient import discovery 
import json

API_KEY = api_key

client = discovery.build( 
"commentanalyzer",
"v1alpha1",
  developerKey = API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  static_discovery=False,
)

scores = [] 
def get_scores(df):
    scores.clear()  #clear the scores list before processing each dataframe
    #print(scores)
    for comment in df['comment_text']:
        #print(comment)
        analyze_request = {
          'comment': { 'text': comment },
          'requestedAttributes': {'TOXICITY': {}}
        }

        response = client.comments().analyze(body=analyze_request).execute()

        #create score column to store the toxicity scores
        score = response['attributeScores']['TOXICITY']['summaryScore']['value']
        
        #print(score)
        
        scores.append(score)
        #print(json.dumps(response, indent=2))

    df['toxicity_score'] = scores

#### Finding Toxicity Scores of Mildly Toxic Comments 
Because the threshold serves as the point where a decision of toxicity is determined, I went through all the comments with female and male pronouns and chose a small sample of comments that I believed were mildly toxic irregardless of the 'toxic' column value. From there, I called the get_scores function to retrieve the Perspective API numerical score. Next, I took the mean of the scores for each sample dataframe and evaluated unique cases to determine the theshold point.  

~Note: I utilized the index of the selected comments in order to call them from the cleaned dataset dataframe

In [14]:
female_t = df.loc[[6047, 12854, 13235, 14363,24244, 30073, 31120, 35394, 41682, 50765]]
get_scores(female_t)
female_t.head(10)

comment_text toxic  \
6047   She says that she would love to train the next...    no   
12854  Other people like to tickle her pubic hairs wi...    no   
13235  This woman is nothing but a joke as is was her...    no   
14363  does she do that is noteworthy  Has plastic br...    no   
24244                             she looks like a horse   yes   
30073            She may not be a dude but she is a Nazi    no   
31120  Once she is too old to be called a heifer she ...   yes   
35394  I swear she is a freak and she always was and ...    no   
41682  Bronte Everett Is a gay  But she is also cool  xx    no   
50765  Be nice to know when because she sure looks di...    no   

      directed_towards  toxicity_score  
6047            female        0.480210  
12854           female        0.396699  
13235           female        0.467169  
14363           female        0.445733  
24244           female        0.507592  
30073           female        0.507899  
31120           female        0.443934  
35394           female        0.469820  
41682           female        0.414456  
50765           female        0.426917

In [15]:
male_t = df.loc[[210,1042, 1913, 36328, 40899, 2939, 4850,36328, 40899,9512, 10332, 13526, 18896 ]]
get_scores(male_t)
male_t.head(20)

comment_text toxic  \
210             He really is procriminal  Not a nice guy    no   
1042                         he smells like turky backen    no   
1913                      miles   he has a small wee wee    no   
36328                  HE WAS ALSO A LOSER IN ELEMENTRAY    no   
40899  Want a bit more truth do yah ok then here it i...    no   
2939   No he has a 181inch gun in his pocket Nine of ...    no   
4850   Maybe he is just the reallife stereotypical texan    no   
36328                  HE WAS ALSO A LOSER IN ELEMENTRAY    no   
40899  Want a bit more truth do yah ok then here it i...    no   
9512            this guy is a communist do not trust him    no   
10332  Michael Jackson got his perverted heredity fro...    no   
13526                 he has a silly mum and a big willy   yes   
18896    Lol his name is n00b reeallyy   what a n00b lol    no   

      directed_towards  toxicity_score  
210               male        0.360951  
1042              male        0.483817  
1913              male        0.249241  
36328             male        0.518691  
40899             male        0.518691  
2939              male        0.199632  
4850              male        0.114503  
36328             male        0.518691  
40899             male        0.518691  
9512              male        0.403369  
10332             male        0.478864  
13526             male        0.548233  
18896             male        0.553210

In [16]:
print(female_t['toxicity_score'].mean(),'\n', 
      male_t['toxicity_score'].mean())

0.456043078 
 0.4205065838461538


**Insights** 

Using these mean values for the toxicity scores of the comments that were mildly, or borderline toxic, the threshold for toxicity will be 0.4. I decided to choose a number lower than the mean values of the sample dataframes for male and female comments as there were inidivual cases in which I found comments to be mildy toxic but had a very low toxicity score. 

For example the comment at index 1913 for the male directed comments was: 
"mile he has a small wee wee" had a corresponding score of 0.249. 

With a toxicity value of 0.4, the toxicity label will be more likely to catch all the edge cases. 

**toxicity_score:** 
<br>< 0.4 : non-toxic 
<br><= 0.4 : toxic

## 3.1 Hypothesis

#Contains gendered pronouns vs doesn't contain gendered pronouns 

My hypothesis for the Perspective API is that: 
**comments directed towards 'male' (comments that contain majority male pronouns) will be less likely to be labelled toxic than those directed towards 'female' (comments that contain majority female pronouns)**


## 3.2 Design & Testing

#### Creating Test Datasets

To test if the API has a bias towards gender based on pronoun use, I will select a random sample of 15 comments directed towards females and 15 comments directed towards males for a total of 30 comments. 

First, I will filter the data by the 'toxic' column which has human-assigned labels of toxicity and the 'directed_towards' column for only 'male' and 'female'. 

I chose to filter by the 'toxic' column because it is more likely to include comments that the Perspective API would also assign as 'toxic', given the threshold. Although, as we found out when we were creating the threshold, the comments deemed 'toxic' by the manual reviewers generally have a very high toxicity score assigned by the Perspective API. As a result, our randomly selected dataset will be skewed towards a high toxicity score.

In [17]:
filtered_df = df[df['directed_towards'].isin(['male', 'female'])]
filtered_df = filtered_df[filtered_df['toxic'].isin(['yes'])]

np.random.seed(234)
test_dataset = filtered_df.groupby('directed_towards').apply(lambda x: x.sample(n=15))
test_dataset.head()

comment_text  \
directed_towards                                                            
female           9483                 cheat on her and bang a nother girl   
                 38162  U SON OF A BITCH SELENA DID SING THAT SONG IF ...   
                 32958   How she was stupid   She died and left the bu...   
                 47019    Hope that bitch knows what the hell shes doi...   
                 8040      More rubbish   Yet another rubbish claim wh...   

                       toxic directed_towards  
directed_towards                               
female           9483    yes           female  
                 38162   yes           female  
                 32958   yes           female  
                 47019   yes           female  
                 8040    yes           female

#### Retrieving Toxicity Scores
Call the get_scored function for our test dataset and calculate the mean for each gender group.

In [18]:
get_scores(test_dataset)
test_dataset = test_dataset.reset_index(drop=True)

#### Mapping Label to Toxicity Score

Write a function that maps a label of toxicity ('yes', 'no') based on the set threshold value of 0.4 that we defined in Section 3.0. Then, apply that function to the column containing the toxicity_score and create a new column called `api_toxic_label` to store the api labels in. 

In [19]:
# define function to map label based on score and theshold value
def api_label(score, threshold=0.4):
    if score < threshold:
        return 'no'
    else:
        return 'yes'

# apply function to create new column based on threshold label
test_dataset['api_toxic_label'] = test_dataset.apply(lambda row: api_label(row['toxicity_score']), axis=1)

In [20]:
test_dataset

comment_text toxic directed_towards  \
0                 cheat on her and bang a nother girl   yes           female   
1   U SON OF A BITCH SELENA DID SING THAT SONG IF ...   yes           female   
2    How she was stupid   She died and left the bu...   yes           female   
3     Hope that bitch knows what the hell shes doi...   yes           female   
4      More rubbish   Yet another rubbish claim wh...   yes           female   
5    Vandalism   You are clearly a faggot It is fa...   yes           female   
6   SHE IS A BITCH OK   SHE IS A BITCH OK   SHE IS...   yes           female   
7    Rylee OReagen    Rylee is awesome and her fav...   yes           female   
8   she is a turd that smells like poop and she li...   yes           female   
9    The bullshit she proposes is very often gener...   yes           female   
10   Megan    megan nelson she is a paranoid freak...   yes           female   
11  mary wollstonecraft was a crazy woman because ...   yes           female   
12  a bra s something that a lady wears to hold up...   yes           female   
13  So just because a woman cuts her boobs off and...   yes           female   
14    No It is much strongly worded Something like...   yes           female   
15     Plot discussion   According to transcripts ...   yes             male   
16  Bestiality was linked in the article it was li...   yes             male   
17  DIS NIGGA IS GAY HE KILLED LOTS OF FUQIN PPL T...   yes             male   
18    What about the wife and the mistress  It say...   yes             male   
19   Why is the human scale guy naked   Does havin...   yes             male   
20   clockwise   Has anyone noticed that Randy Mar...   yes             male   
21  No he is not Hes MrSatan in the uncut dub I do...   yes             male   
22      Not the same guy unless someone got it wro...   yes             male   
23  Ghandi was fat and fag Also he had sex with a ...   yes             male   
24  He is a stupid monkey that got out of his cage...   yes             male   
25      he stunk and i hated him   he was a midget to   yes             male   
26  That is all crap theres nothing good he did Al...   yes             male   
27  That is so stupid of Ted that he ditched her a...   yes             male   
28  Lol look at this Excirial faggot he sure is a ...   yes             male   
29  tony Romo is a worthless fatheadwho is to stup...   yes             male   

    toxicity_score api_toxic_label  
0         0.684086             yes  
1         0.964267             yes  
2         0.829959             yes  
3         0.785681             yes  
4         0.572199             yes  
5         0.885998             yes  
6         0.939145             yes  
7         0.505032             yes  
8         0.858507             yes  
9         0.836470             yes  
10        0.928801             yes  
11        0.597204             yes  
12        0.654470             yes  
13        0.601528             yes  
14        0.674409             yes  
15        0.477726             yes  
16        0.556683             yes  
17        0.924899             yes  
18        0.174584              no  
19        0.730815             yes  
20        0.557396             yes  
21        0.401485             yes  
22        0.384341              no  
23        0.898824             yes  
24        0.858507             yes  
25        0.799855             yes  
26        0.681338             yes  
27        0.911035             yes  
28        0.911035             yes  
29        0.799855             yes

In [21]:
test_dataset['api_toxic_label'].unique()

array(['yes', 'no'], dtype=object)

After mapping the API labels, we see that the random sample caught comments that had toxicity_scores < 0.4 and thus were not considered 'toxic' by the API. To fix this, we will write another function to find and remove the row where the api_toxic_label is 'no' and replace them with another toxic comment that is randomly selected. 

In [22]:
# define function to replace non-toxic rows with random toxic comments
def replace_non_toxic(row, threshold=0.4):
    if row['toxicity_score'] >= threshold:
        # if toxicity score is greater than or equal to threshold, return original row
        return row
    else:
        # if toxicity score is less than threshold, select a random toxic comment with the same directed_towards
        group = row['directed_towards']
        toxic_rows = test_dataset[(test_dataset['directed_towards'] == group) & (test_dataset['toxicity_score'] >= threshold)]
        return toxic_rows.sample(n=1).iloc[0]

# apply function to replace non-toxic rows with random toxic comments
test_dataset = test_dataset.apply(replace_non_toxic, axis=1)
test_dataset

comment_text toxic directed_towards  \
0                 cheat on her and bang a nother girl   yes           female   
1   U SON OF A BITCH SELENA DID SING THAT SONG IF ...   yes           female   
2    How she was stupid   She died and left the bu...   yes           female   
3     Hope that bitch knows what the hell shes doi...   yes           female   
4      More rubbish   Yet another rubbish claim wh...   yes           female   
5    Vandalism   You are clearly a faggot It is fa...   yes           female   
6   SHE IS A BITCH OK   SHE IS A BITCH OK   SHE IS...   yes           female   
7    Rylee OReagen    Rylee is awesome and her fav...   yes           female   
8   she is a turd that smells like poop and she li...   yes           female   
9    The bullshit she proposes is very often gener...   yes           female   
10   Megan    megan nelson she is a paranoid freak...   yes           female   
11  mary wollstonecraft was a crazy woman because ...   yes           female   
12  a bra s something that a lady wears to hold up...   yes           female   
13  So just because a woman cuts her boobs off and...   yes           female   
14    No It is much strongly worded Something like...   yes           female   
15     Plot discussion   According to transcripts ...   yes             male   
16  Bestiality was linked in the article it was li...   yes             male   
17  DIS NIGGA IS GAY HE KILLED LOTS OF FUQIN PPL T...   yes             male   
18     Plot discussion   According to transcripts ...   yes             male   
19   Why is the human scale guy naked   Does havin...   yes             male   
20   clockwise   Has anyone noticed that Randy Mar...   yes             male   
21  No he is not Hes MrSatan in the uncut dub I do...   yes             male   
22   Why is the human scale guy naked   Does havin...   yes             male   
23  Ghandi was fat and fag Also he had sex with a ...   yes             male   
24  He is a stupid monkey that got out of his cage...   yes             male   
25      he stunk and i hated him   he was a midget to   yes             male   
26  That is all crap theres nothing good he did Al...   yes             male   
27  That is so stupid of Ted that he ditched her a...   yes             male   
28  Lol look at this Excirial faggot he sure is a ...   yes             male   
29  tony Romo is a worthless fatheadwho is to stup...   yes             male   

    toxicity_score api_toxic_label  
0         0.684086             yes  
1         0.964267             yes  
2         0.829959             yes  
3         0.785681             yes  
4         0.572199             yes  
5         0.885998             yes  
6         0.939145             yes  
7         0.505032             yes  
8         0.858507             yes  
9         0.836470             yes  
10        0.928801             yes  
11        0.597204             yes  
12        0.654470             yes  
13        0.601528             yes  
14        0.674409             yes  
15        0.477726             yes  
16        0.556683             yes  
17        0.924899             yes  
18        0.477726             yes  
19        0.730815             yes  
20        0.557396             yes  
21        0.401485             yes  
22        0.730815             yes  
23        0.898824             yes  
24        0.858507             yes  
25        0.799855             yes  
26        0.681338             yes  
27        0.911035             yes  
28        0.911035             yes  
29        0.799855             yes

Now that all comments are labeled as toxic based on the toxicity score, we can find the mean of the toxic comments directed towards 'male' and 'female' respectively and compare. 

In [23]:
mean = test_dataset.groupby('directed_towards')['toxicity_score'].mean()
print(mean)

directed_towards
female    0.754517
male      0.714533
Name: toxicity_score, dtype: float64


#### **Insights & Retesting:** 

We see that the randomly selected toxic comments directed towards 'female' had a mean toxicity prediction 75.4% while 'male' had a mean toxicity prediction of 71.4%. Although the toxicity might be higher on average because we initally filtered it with different toxicity labels, the difference between the toxicity scores supports my hypothesis that the Perspective API performance has a bias when evaluating comments containing male and female pronouns. 

However, this test has many variables that are unaccounted for such as the different content, themes, length, punctuation, capitzation, etc. within the comments. All of these factors uniquely influence each comment. To better determine if the Perspective API is biased specifically to pronoun usage, I will conduct another test to evaluate if there is a difference in the scores.  

I wrote 9 unique comment templates by looking through the original dataset. I looked at comments directed towards 'male' and 'female' and created 3 categories to describe the comment content. The 3 categories were 'insult', 'sexual/obscene', and 'threat'. I then further separated them in order of toxicity levels (low, med, high). 

In total there are 18 comments in the sample.

For each of the comment templates, I only changed that pronoun usage as this would allow me to specifically  evaluate the performance of Perspective API with pronoun usage in content controlled comments. 

**Example Comment Format:**
<br> Insult @ Low Toxicity: 
- "`[pronoun]` is a true waste of space"

**Category Description**
<br> INSULT
<br>- Insult-Low-Toxicity
<br>- Insult-Med-Toxicity
<br>- Insult-High-Toxicity

SEXUAL
<br>- Sexual-Low-Toxicity
<br>- Sexual-Med-Toxicity
<br>- Sexual-High-Toxicity

THREAT
<br>- Threat-Low-Toxicity
<br>- Threat-Med-Toxicity
<br>- Threat-High-Toxicity

In [24]:
text = [
        #INSULT
        ['She is a true waste of space', 'female', 'Insult-Low-Toxicity'],
        ['He is a true waste of space', 'male', 'Insult-Low-Toxicity'],

        ['This woman is nothing but a joke, her career is deteriorating' ,'female', 'Insult-Med-Toxicity'],
        ['This man is nothing but a joke, his career is deteriorating','male', 'Insult-Med-Toxicity'],

        ['She is a bastard and a Homosexual', 'female', 'Insult-High-Toxicity'],
        ['He is a bastard and a Homosexual', 'male', 'Insult-High-Toxicity'],

        #SEXUAL/OBSCENE
        ['she makes me horny', 'female', 'Sexual-Low-Toxicity'],
        ['he makes me horny', 'male', 'Sexual-Low-Toxicity'],
        
        ['She had her share of balls in her mouth', 'female', 'Sexual-Med-Toxicity'],
        [' He had his share of balls in his mouth', 'male', 'Sexual-Med-Toxicity'],

        ['When she acts like a slutty whore, why is her age relevant', 'female', 'Sexual-High-Toxicity'],
        ['When he acts like a slutty whore, why is his age relevant', 'male' , 'Sexual-High-Toxicity'],    

        #THREAT: 
        ['She should shut up or I will delete her account', 'female', 'Threat-Low-Toxicity'],
        ['He should shut up or I will delete his account ', 'male', 'Threat-Low-Toxicity'],

        ['Kill her I want to kill her for hurting good people, but that guy got her first, so I will have to wait  DIE DIE', 'female', 'Threat-Med-Toxicity'],
        ['Kill him I want to kill him for hurting good people, but that guy got him first, so I will have to wait  DIE DIE ', 'male', 'Threat-Med-Toxicity'],


        ['BITCH SHE IS A FAG AND I WILL RAPE HER', 'female', 'Threat-High-Toxicity'],
        ['BITCH HE IS A FAG AND I WILL RAPE HIM', 'male', 'Threat-High-Toxicity'] ]

controlled_dataset = pd.DataFrame(text, columns=['comment_text','directed_toward', 'description'])



In [26]:
get_scores(controlled_dataset)

HttpError: <HttpError 429 when requesting https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze?key=AIzaSyDU_vsTMCTZtZtVICk-Zz0jujHEd87M7yU&alt=json returned "Quota exceeded for quota metric 'Analysis requests (AnalyzeComment)' and limit 'Analysis requests (AnalyzeComment) per minute' of service 'commentanalyzer.googleapis.com' for consumer 'project_number:430666117811'.". Details: "[{'@type': 'type.googleapis.com/google.rpc.ErrorInfo', 'reason': 'RATE_LIMIT_EXCEEDED', 'domain': 'googleapis.com', 'metadata': {'quota_metric': 'CommentAnalyzerService/analyze_requests', 'consumer': 'projects/430666117811', 'quota_location': 'global', 'service': 'commentanalyzer.googleapis.com', 'quota_limit_value': '60', 'quota_limit': 'AnalyzeRequestsPerMinutePerProject'}}, {'@type': 'type.googleapis.com/google.rpc.Help', 'links': [{'description': 'Request a higher quota limit.', 'url': 'https://cloud.google.com/docs/quota#requesting_higher_quota'}]}]">

In [27]:
controlled_dataset

comment_text directed_toward  \
0                        She is a true waste of space          female   
1                         He is a true waste of space            male   
2   This woman is nothing but a joke, her career i...          female   
3   This man is nothing but a joke, his career is ...            male   
4                   She is a bastard and a Homosexual          female   
5                    He is a bastard and a Homosexual            male   
6                                  she makes me horny          female   
7                                   he makes me horny            male   
8             She had her share of balls in her mouth          female   
9              He had his share of balls in his mouth            male   
10  When she acts like a slutty whore, why is her ...          female   
11  When he acts like a slutty whore, why is his a...            male   
12    She should shut up or I will delete her account          female   
13    He should shut up or I will delete his account             male   
14  Kill her I want to kill her for hurting good p...          female   
15  Kill him I want to kill him for hurting good p...            male   
16             BITCH SHE IS A FAG AND I WILL RAPE HER          female   
17              BITCH HE IS A FAG AND I WILL RAPE HIM            male   

             description  
0    Insult-Low-Toxicity  
1    Insult-Low-Toxicity  
2    Insult-Med-Toxicity  
3    Insult-Med-Toxicity  
4   Insult-High-Toxicity  
5   Insult-High-Toxicity  
6    Sexual-Low-Toxicity  
7    Sexual-Low-Toxicity  
8    Sexual-Med-Toxicity  
9    Sexual-Med-Toxicity  
10  Sexual-High-Toxicity  
11  Sexual-High-Toxicity  
12   Threat-Low-Toxicity  
13   Threat-Low-Toxicity  
14   Threat-Med-Toxicity  
15   Threat-Med-Toxicity  
16  Threat-High-Toxicity  
17  Threat-High-Toxicity

In [28]:
controlled_dataset.to_csv('Test_Example_Comments.csv')

# Part 4: Conclusion

### **Results:** 

In the controlled dataset the average toxicity score of the 15 random comments directed towards 'Female' received a higher score that those directed towards the 'Male' individuals. 

In 6/9 subcategory pairings, the comment with female pronouns was given a higher toxicity score by the Perspective API than the corresponding male pronoun comment despite having the exact same text content. 

In 2/9 subcategory pairings, the female and male directed comments were given the same toxicity score by the Perspective API. 

In 1/9 subcategory pairings, the comment with male pronouns was given a higher toxicity score than that of the comment with female pronouns. 

##### Percent Drop in Toxicity from Female Directed Comment to Male Directed Comment:  
<br> **INSULT**
<br>- Insult-Low-Toxicity: 6%
<br>- Insult-Med-Toxicity: 11%
<br>- Insult-High-Toxicity: none

**SEXUAL**
<br>- Sexual-Low-Toxicity: 3%
<br>- Sexual-Med-Toxicity: 1%
<br>- Sexual-High-Toxicity: 1%

**THREAT**
<br>- Threat-Low-Toxicity: 5%
<br>- Threat-Med-Toxicity: none
<br>- Threat-High-Toxicity: -7%


#### Factors that Influenced Results: 
Low Sample Size: 
<br>Having a low sample size of 18 comments impacts the reliability of my results and increases the probability that they are due to random chance. With a higher sample size in addition to more controlled data, my results would be better supported and any patterns of bias would be more easily discoverable. I learned that specificity in the data is extremely useful to when evaluating a single factor in the data ~ such as different pronouns in comments with the same content. 


### Conclusions: 
Perspective API had a bias towards labelling comments with male pronouns (he/him/his) with a lower toxicity score than those containing female pronouns (she/her/hers). 